### ISSUE TITLE
Where are box / cls / DFL losses defined and wired into the training loop in Ultralytics YOLO?

### ISSUE THREAD

**Issue by djikstra0501** [ultralytics/ultralytics#23432]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi, I’m trying to customize the detection training losses in the Ultralytics YOLO repo (box loss, cls loss, and DFL loss), specifically on YOLOv11 variants. I already found the loss implementations in loss.py (and there are multiple loss classes/functions there), but I can’t see where any of those loss classes/functions are actually called from the model head or the Trainer. This makes me confuse about where the actual loss used for backprop is computed and how a specific loss variant is selected.
What I’m trying to understand:
1. Where is loss computed, exactly?
    •	Which function/method is responsible for computing the training loss from (preds, batch)?
    •	Where is the call site that triggers loss computation during training (the exact file + method name)?
2. How does Ultralytics choose which loss variant to use?
    •	For detection, how are the concrete implementations for:
        o	box loss
        o	cls loss
        o	DFL loss
        o	selected at runtime?
    •	Is this determined by:
        o	the model/task type (detect/seg/pose/obb),
        o	the head type,
        o	model args/hyperparams,
        o	a criterion object attached to the model,
        o	or something else?
3. Call flow / data flow
    •	What is the call chain from Trainer.train() → model(...) → loss?
    •	Where do target preprocessing steps happen (assignment/matching, target building, decoding) before box/cls/DFL are computed?
4. Best way to swap in custom losses
    •	What’s the cleanest/expected hook to replace all three (box/cls/DFL) while keeping everything else intact?
    •	If the repo uses a criterion object, where is it created and how do I override it without breaking other tasks?
I want to ensure I’m overriding the actual loss used in backprop. Thank you !

**Comment by UltralyticsAssistant** [comment]
👋 Hello @djikstra0501, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (like customizing detection losses), please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). To help an Ultralytics engineer quickly point you to the exact call sites and wiring, please also include:-Your `ultralytics` version (`pip show ultralytics`)-Your OS + Python + PyTorch versions-Your exact model name/config and task (detect/seg/pose/obb)-A short snippet showing how you start training (Python or CLI)📌 This is an automated response, and an Ultralytics engineer will also assist soon 🤝🙂

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
It's called in the `model.loss()` method where it dynamically selects the loss function in `init_criterion()`

https://github.com/ultralytics/ultralytics/blob/74b2b91e5c4420af47875801b8e13333ab7026ac/ultralytics/nn/tasks.py#L321

**Comment by krgkaushik** [comment]
Hi @djikstra0501,

Here is the quick breakdown of the flow and how to override it:

1. Where is loss computed?

The Orchestrator: ultralytics/nn/tasks.py in BaseModel.loss().

The Implementation: ultralytics/utils/loss.py class v8DetectionLoss.

The Trigger: Trainer calls model.loss(batch, preds) during the loop.

2. How is it selected?

It is determined by the model's head type (e.g., Detect, Segment).

BaseModel.init_criterion() inspects the last layer. If it sees a Detect head, it instantiates v8DetectionLoss.

3. How to swap it (The cleanest way) You don't need to fork the code. You can monkey-patch the criterion attribute on your model instance before calling .train(). 

```from ultralytics import YOLO
from ultralytics.utils.loss import v8DetectionLoss

class MyCustomLoss(v8DetectionLoss):
    def __call__(self, preds, batch):
        # Implement your custom logic here
        print("DEBUG: Custom loss running")
        return super().__call__(preds, batch)

# Load model
model = YOLO("yolo11n.pt")

# Force-init the model so the head/criterion are built
model.model.args = model.overrides 

# Swap the loss function
model.model.criterion = MyCustomLoss(model.model)

# Train (it will now use your custom class)
model.train(data="coco8.yaml", epochs=1)```

**Comment by djikstra0501** [comment]
Thanks for the information Now i can do some custom loss

**Comment by glenn-jocher** [comment]
Sounds good—when you validate your custom loss, a quick sanity check is to log once inside your overridden `__call__` and confirm `model.model.criterion` is your class; also note that some variants wrap detection loss (e.g. end-to-end uses [`E2ELoss`](https://docs.ultralytics.com/reference/utils/loss/#class-ultralytics.utils.loss.E2ELoss) and prompt-based uses [`TVPDetectLoss`](https://docs.ultralytics.com/reference/utils/loss/#class-ultralytics.utils.loss.TVPDetectLoss)), so make sure you’re patching the wrapper that `model.loss()` actually instantiates (see [`YOLOEModel.loss()`](https://docs.ultralytics.com/reference/nn/tasks/#ultralytics.nn.tasks.YOLOEModel.loss)); if you hit anything unexpected, please open a new issue with the full traceback and `yolo checks`.
